solarkraft/test/e2e/verify_js_example_xycloans.ts

```diff
@@ -0,0 +1,134 @@
+/**
+ * This is an example of how to write a Solarkraft runtime monitor in
+ * TypeScript.
+ *
+ * This is a monitor for the pool contract from Xycloans:
+ * https://github.com/xycloo/xycloans/tree/e0663724eebf2ff52d59c836b697e97cd9f629a8/pool/src
+ *
+ * The test targets a historical commit known to contain a (now fixed) vulnerability,
+ * where the fee per share is subject to a rounding error, found by OtterSec:
+ * https://github.com/xycloo/xycloans/blob/main/audits/osec.pdf
+ * https://github.com/xycloo/xycloans/commit/b26e0124ab0b84c1be25c636492671f106653f09
+ * 
+ * The monitor checks that the fee per share is calculated correctly and that
+ * the rewards are distributed correctly.
+ * 
+ * @author Thomas Pani, 2024
+ */
+import {
+    every,
+    Env,
+    EnvRevert,
+    SolarkraftJsMonitor,
+    tokenReceived,
+    tokenTransferred,
+} from '../../src/verify_js.js'
+import { OrderedMap } from 'immutable'
+import { Divide, ROUNDING_MODE } from 'big-round'
+
+const divideCeil = Divide(ROUNDING_MODE.DIRECTED_TOWARDS_POSITIVE_INFINITY)
+const divideFloor = Divide(ROUNDING_MODE.DIRECTED_TOWARDS_NEGATIVE_INFINITY)
+
+const STROOP = 10_000_000n
+
+const XLM_TOKEN_SAC_TESTNET = 'CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC'
+
+/* the monitor */
+
+class XycloansPoolMonitor extends SolarkraftJsMonitor {
+    // shares per address
+    private shares = OrderedMap<string, bigint>()
+    // fee per share for the entire pool, in stroops
+    private fee_per_share_universal = 0n
+
+    // total number of shares
+    private total_shares = () => this.shares.reduce((acc, x) => acc + x, 0n)
+
+    initialize(token: string) {
+        this.reverts_if((env: EnvRevert) => {
+            return env.storage().instance().has('TokenId')
+        })
+
+        this.succeeds_with((env: Env) => {
+            return every(
+                // check that `TokenId` is set to the XLM SAC
+                token == XLM_TOKEN_SAC_TESTNET,
+                env.storage().instance().get('TokenId') == token
+            )
+        })
+    }
+
+    deposit(from: string, amount: bigint) {
+        this.succeeds_with((env: Env) => {
+            const token = env.storage().instance().get('TokenId')
+
+            // side-effects: we track deposits locally to compute rewards later
+            this.shares = this.shares.update(from, 0n, (x) => x + amount)
```

Comment:
This looks like a very fragile solution to me. What if we have two `succeeds_with` conditions that both update `this.shares`?

Reply:
We could move it outside the `succeeds_with`. WDYT?

Reply:
yes, I believe it's how it should be done conceptually. Kind of like one update per tx.

Reply:
Uh, but here's the thing: we're passing `Env` to `succeeds_with` not to the containing method. However, the second side-effect below accesses storage via `env`.

We're doing this because `reverts_if` and `succeeds_with` receive different `Env`s: `reverts_if` only has access to the pre-state, while `succeeds_with` receives pre- and post-states.

Reply:
I will think about it a bit more, but maybe we need to leave this as a future design consideration.